### PR TITLE
Update service ps to be an API test

### PR DESCRIPTION
**Pull request by dnephin**
Removed cli tests (which were testing CLI behaviour) and replaced with an API test

Also update `LoadBusybox()` to use the API instead of the cli

### CODE REVIEW

**Review by thaJeztah — APPROVED** [review]
LGTM

**Comment by aaronlehmann** [comment]
LGTM

**Comment by thaJeztah** [comment]
Given that this test is testing CLI functionality, but does not catch the bug in https://github.com/moby/moby/pull/32800

Is there an actual need to merge this? https://github.com/moby/moby/pull/32800#issuecomment-308605830 suggests rewriting to an API test (which won't do any testing here, as it's only consecutive calls to the same endpoint)

Shall we close this PR and remove the test?

**Comment by dnephin** [comment]
If we have other test coverage for this endpoint I would agree we can just remove this test entirely.

**Comment by dnephin** [comment]
It looks like we don't have any other tests for this endpoint, so this will need to be converted to api tests.

**Comment by dnephin** [comment]
OK, this PR has been updated accordingly

**Comment by aaronlehmann** [comment]
Is there an equivalent test in `cli` to the one that is being removed?

**Comment by dnephin** [comment]
The test should exist in e2e soon. 

There is no equivalent in `docker/cli` right now. We're missing a lot of coverage in that repo. I opened https://github.com/docker/cli/issues/201 as a tracking ticket for test cases we remove from this repo.

I don't think we should wait for those tests to exist before merging this and similar PRs. This test fails on the jenkins integration job because of changes to the CLI . Test coverage of the code in this repo remains the same after this change. We can't rely on tests in this repo to cover behaviour in another repo.

**Comment by aaronlehmann** [comment]
> I don't think we should wait for those tests to exist before merging this and similar PRs.

I'm afraid I disagree.

> This test fails on the jenkins integration job because of changes to the CLI.

So let's fix it instead of removing it.

> We can't rely on tests in this repo to cover behaviour in another repo.

We need some way of making sure that the CLI works against the daemon. Currently it's impossible to add tests that verify this, and removing existing tests will make the situation even worse.

**Comment by dnephin** [comment]
> So let's fix it instead of removing it.

That's **exactly** what I did here! The fix is to make it an API test.
cc @andrewhsu since we've been talking about this.

> We need some way of making sure that the CLI works against the daemon. Currently it's impossible to add tests that verify this, and removing existing tests will make the situation even worse.

Yes we do. It is the responsibility of `docker/cli` to make sure it works with the daemon. 

The daemon should only be responsible for preventing regressions. Using a client binary is not the right way to do that, it hides regressions. We need to write API tests that check expected responses.

That said, the API test I wrote is terrible. I need to update it to expect an exact struct and raw bytes.

**Comment by aaronlehmann** [comment]
> That's **exactly** what I did here! The fix is to make it an API test.

The old test checks the behavior of `service ps` with prefixes and multiple arguments. The replacement tests the `ServiceInspect` API method with a single full-length ID. I don't see how one is a substitute for the other.

**Comment by dnephin** [comment]
> The old test checks the behavior of service ps with prefixes and multiple arguments.

That logic is in `docker/cli`. We can not test it here. The only code in this repo that is being exercised by those tests is this one API call.

**Comment by cpuguy83** [comment]
I'm not comfortable removing the CLI test until there is a valid replacement in docker/cli.

**Comment by dnephin** [comment]
I was wrong, there are actually unit tests covering this: https://github.com/docker/cli/blob/master/cli/command/service/ps_test.go

What value does this test provide? It's testing with a arbitrary frozen version of the CLI. We already know [this test is failing](https://jenkins.dockerproject.org/job/docker-integration-tests/) with the latest version of the cli. 

My first change was to update the test, but @thaJeztah correctly pointed out that it doesn't make sense to keep this test. It's not actually testing anything.

**Review by thaJeztah — APPROVED** [review]
LGTM

ping @cpuguy83 good to go?

**Comment by dnephin** [comment]
This isn't ready yet. I want to update the test to assert the entire struct matches the expect struct, instead of just comparing ID.

**Comment by dnephin** [comment]
I've updated the test to compare the entire service instead of just the ID. This is ready for review again.

**Review by thaJeztah — APPROVED** [review]
still LGTM

**Comment by mlaventure** [comment]
@dnephin this need a rebase

**Comment by dnephin** [comment]
Thanks, rebased

**Comment by thaJeztah** [comment]
Failures look related;

```
22:40:24 ----------------------------------------------------------------------
22:40:24 FAIL: docker_api_swarm_service_test.go:682: DockerSwarmSuite.TestAPISwarmServicesInspectFull
22:40:24 
22:40:24 [d511d16ba1c2f] waiting for daemon to start
22:40:24 [d511d16ba1c2f] daemon started
22:40:24 
22:40:24 docker_api_swarm_service_test.go:697:
22:40:24     waitAndAssert(c, defaultReconciliationTimeout, d.CheckActiveContainerCount, checker.Equals, int(instances))
22:40:24 docker_utils_test.go:471:
22:40:24     c.Assert(v, checker, args...)
22:40:24 ... obtained int = 0
22:40:24 ... expected int = 2
22:40:24 
22:40:24 [d511d16ba1c2f] exiting daemon
22:40:49 
```

**Comment by dnephin** [comment]
Hmm, yes it does, but I can't reproduce locally.

**Comment by dnephin** [comment]
I don't see any errors in `bundles/17.06.0-dev/test-integration-cli/d326d42de1f20/docker.log`

**Comment by thaJeztah** [comment]
ping @dperny @vdemeester could you have a look at this one; perhaps you have an idea

**Comment by dperny** [comment]
Something changed, I'm only seeing the PPC test failure which is unrelated.

**Comment by thaJeztah** [comment]
oh! that's cool, I thought, well, let's try once more; looks like that resolved it, must've been fixed elsewhere then

**Comment by dnephin** [comment]
This has been updated again to move the API test to the new `integration/` suite

File: integration/service/main_test.go [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+func TestMain(m *testing.M) {
+	var err error
+	testEnv, err = environment.New()
+	if err != nil {
+		fmt.Println(err)
+		os.Exit(1)
+	}
+
+	// TODO: replace this with `testEnv.Print()` to print the full env
```
Comment (vdemeester) [inline integration/service/main_test.go]: ❤️

**Review by vdemeester — APPROVED** [review]
LGTM 🐯

**Pull request merged**